src/httpauth/digest.c

```diff
@@ -591,3 +613,405 @@ int httpauth_digest_chall_request_full(struct httpauth_digest_chall_req **preq,
+{
+	struct httpauth_digest_enc_resp *resp = NULL;
+	uint32_t cnonce = rand_u32();
+	int err = 0, n = 0;
+
+	if (!presp || !chall || !method || !uri || !user || !passwd)
+		return EINVAL;
+
+	resp = mem_zalloc(sizeof(*resp), httpauth_digest_response_destructor);
+	if (!resp) {
+		return ENOMEM;
+	}
+
+	resp->cnonce = mem_zalloc(CNONCE_NC_LENGTH, NULL);
+	resp->nc = mem_zalloc(CNONCE_NC_LENGTH, NULL);
+	if (!resp->cnonce || !resp->nc) {
+		err = ENOMEM;
+		goto out;
+	}
+
+	/* copy fields */
+	err = pl_strdup(&resp->realm, &chall->realm);
+	err |= pl_strdup(&resp->nonce, &chall->nonce);
+	err |= pl_strdup(&resp->opaque, &chall->opaque);
+	if (err) {
+		goto out;
+	}
+
+	/* userhash supported by server */
+	if (userhash && (pl_strcasecmp(&chall->userhash, "true") == 0))
+		resp->userhash = true;
+
+	/* only allowed qop Nothing, "auth" or "auth-int" */
+	if (str_isset(qop) && (str_casecmp(qop, "auth")) &&
+		(str_casecmp(qop, "auth-int"))) {
+		err = EPROTONOSUPPORT;
+		goto out;
+	}
+
+	/* qop supported by server */
+	if (pl_isset(&chall->qop) && str_isset(qop) &&
+		pl_strstr(&chall->qop, qop)) {
+		err = str_dup(&resp->qop, qop);
+		if (err)
+			goto out;
+	}
+
+	/* only allowed charset Nothing or "UTF-8" */
+	if (str_isset(charset) && str_casecmp(charset, "UTF-8")) {
+		err = EPROTONOSUPPORT;
+		goto out;
+	}
+
+	/* charset supported by server */
+	if (pl_isset(&chall->charset) && str_isset(charset) &&
+		pl_strstr(&chall->charset, charset) == 0) {
+		err = str_dup(&resp->charset, charset);
+		if (err)
+			goto out;
+	}
+
+	err = str_dup(&resp->uri, uri);
+	if (err)
+		goto out;
+
+	n = re_snprintf(resp->cnonce, CNONCE_NC_LENGTH, "%08x", cnonce);
+	if (n == -1 || n != CNONCE_NC_LENGTH -1) {
+		err = ERANGE;
+		goto out;
+	}
+
+	n = re_snprintf(resp->nc, CNONCE_NC_LENGTH, "%08x", nc++);
+	if (n == -1 || n != CNONCE_NC_LENGTH -1) {
+		err = ERANGE;
+		goto out;
+	}
+
+	if (pl_strstr(&chall->algorithm, "SHA256-sess")) {
+		resp->hash_function = &sha256;
+		resp->hash_length = SHA256_DIGEST_LENGTH;
+		err = str_dup(&resp->algorithm, "SHA256-sess");
+	}
+	else if (pl_strstr(&chall->algorithm, "SHA256")) {
+		resp->hash_function = &sha256;
+		resp->hash_length = SHA256_DIGEST_LENGTH;
+		err = str_dup(&resp->algorithm, "SHA256");
+	}
+	else if (pl_strstr(&chall->algorithm, "SHA1-sess")) {
+		resp->hash_function = &sha1;
+		resp->hash_length = SHA_DIGEST_LENGTH;
+		err = str_dup(&resp->algorithm, "SHA1-sess");
+	}
+	else if (pl_strstr(&chall->algorithm, "SHA1")) {
+		resp->hash_function = &sha1;
+		resp->hash_length = SHA_DIGEST_LENGTH;
+		err = str_dup(&resp->algorithm, "SHA1");
+	}
+	else if (pl_strstr(&chall->algorithm, "MD5-sess")) {
+		resp->hash_function = &md5;
+		resp->hash_length = MD5_SIZE;
+		err = str_dup(&resp->algorithm, "MD5-sess");
+	}
+	else if (!pl_isset(&chall->algorithm) ||
+		pl_strstr(&chall->algorithm, "MD5")) {
+		resp->hash_function = &md5;
+		resp->hash_length = MD5_SIZE;
+		err = str_dup(&resp->algorithm, "MD5");
```

Comment:
We should consider how to deal with weak crypto such as MD5 or SHA1.

Reply:
for now it should be enough with MD5 and SHA256.

I am wondering if we should add a global macro ENABLE_WEAK_CIPHERS that we can put around
the MD5 and SHA1 code... ?

Reply:
Disabling the MD5 via an macro is quite risky. It might break a lot of digest authentication implementations.
For SHA1 on the other hand it would be a good idea since the SHA2 family should be used instead.

I will remove the SHA1 algorithm for now.